features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/networksettled/ViewNetworkSettledMetric.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.rum.internal.metric.networksettled
+
+import com.datadog.android.api.InternalLogger
+import java.util.PriorityQueue
+import java.util.concurrent.TimeUnit
+
+internal class ViewNetworkSettledMetric(
+    requestIdentificationStrategy: NetworkSettledRequestIdentificationStrategy,
+    private val internalLogger: InternalLogger,
+    private val requestValidator: RequestValidator = resolveRequestValidator(requestIdentificationStrategy)
+) {
+
+    private val resourceStartedTimestamps = HashMap<String, HashMap<String, Long>>()
+    private val resourceStoppedTimestamps = HashMap<String, PriorityQueue<Long>>()
+    private val viewCreatedTimestampsByBViewId = HashMap<String, Long>()
+
+    fun viewWasCreated(viewId: String, eventTimestampInNanos: Long) {
+        viewCreatedTimestampsByBViewId[viewId] = eventTimestampInNanos
+        purge()
+    }
+
+    fun resourceWasStarted(context: NetworkSettledResourceContext, eventTimestampInNanos: Long) {
+        // check if the resource was is a network settled valid resource
+        if (requestValidator.validate(context, viewCreatedTimestampsByBViewId)) {
+            // check if we have a view created entry for this resource
+            val viewCreatedTimestamp = viewCreatedTimestampsByBViewId[context.viewId]
+            if (viewCreatedTimestamp != null &&
+                System.nanoTime() - viewCreatedTimestamp < RESOURCE_START_MAX_THRESHOLD_NS
+            ) {
+                resourceStartedTimestamps.getOrPut(context.viewId) { HashMap() }[context.resourceId] =
+                    eventTimestampInNanos
+            }
+        }
+        purge()
+    }
+
+    fun resourceWasStopped(context: NetworkSettledResourceContext, eventTimestampInNanos: Long) {
+        // check if we have a start timestamp for this resource
+        val startTimestamp = resourceStartedTimestamps[context.viewId]?.get(context.resourceId)
+        if (startTimestamp != null) {
+            val resourceSettleDuration = eventTimestampInNanos - startTimestamp
+            resourceStoppedTimestamps.getOrPut(context.viewId) { PriorityQueue() }.add(resourceSettleDuration)
```

Comment:
question/ Do we need a `PriorityQueue()`? At the end, all we're looking for is the `eventTimestampInNanos` of the last stopped resource which can be stored on single property. As long as we can assume that events arrive in order, the same can be achieved without using any data structures:
```kotlin
private var numberOfPendingResources: Int = 0
private var timestampOfLastStoppedResourceNs: Long?

fun resourceWasStarted() {
   numberOfPendingResources += 1
}

fun resourceWasStopped(eventTimestampInNanos: Long) {
   timestampOfLastStoppedResourceNs = eventTimestampInNanos
   numberOfPendingResources -= 1
}
```

Reply:
you need a `MaxHeap` there (in Java that's a PriorityQueue) in order to make sure you have everything ordered as you cannot guarantee the resources are going to be stopped in the same order. So you need to make sure that the last one is the max. In the end you do not need that if you prefer to do the `sort` when you compute and pick the highest one. For me I prefer this because it is a bit more efficient.